internal/db/schema/migrations/oss/postgres/22/01_session_valid_state.up.sql

```diff
@@ -0,0 +1,93 @@
+begin;
+
+-- Remove invalid (states with no prior end time) rows
+delete from session_state where previous_end_time is null and state != 'pending';
+
+-- In the case of dupes that are otherwise valid, identify first valid session
+create temp table session_state_delete
+as select session_id, min(start_time) as st_t from session_state ou
+   where (select count(*) from session_state inr
+          where inr.session_id = ou.session_id and inr.state = ou.state) > 1
+   group by session_id;
+
+-- Remove all session rows after first valid session
+DELETE FROM session_state t1
+    using session_state_delete t2 where t1.session_id =t2.session_id and t1.start_time > t2.st_t;
+
+drop table session_state_delete;
+
+-- Close all open session_connections
+update session_connection
+set closed_reason='canceled' where closed_reason is null;
+
+update session
+set termination_reason='canceled' where termination_reason is null;
+
+-- Migration
+
+-- session_valid_state table creation and related constraints
+create table session_valid_state(
+    prior_state text
+        references session_state_enm(name)
+            on delete restrict
+            on update cascade,
+        constraint valid_prior_states
+        check (
+                prior_state in ('pending', 'active', 'canceling')
+            ),
+    current_state text
+        references session_state_enm(name)
+            on delete restrict
+            on update cascade,
+        constraint valid_current_states
+        check (
+            current_state in ('pending', 'active', 'canceling', 'terminated')
+        ),
+    primary key (prior_state, current_state)
+);
+
+insert into session_valid_state (prior_state, current_state)
+values
+    ('pending','pending'),
+    ('pending','active'),
+    ('pending','terminated'),
+    ('pending','canceling'),
+    ('active','canceling'),
+    ('active','terminated'),
+    ('canceling','terminated');
+
+alter table session_state
+    add column prior_state text not null default 'pending'
+        references session_state_enm(name)
+        on delete restrict
+        on update cascade;
+alter table session_state
+    add foreign key (prior_state, state)
+      references session_valid_state (prior_state,current_state);
+alter table session_state
+    add unique (session_id, state);
+
+create or replace function
```

Comment:
We have not always been consistent with this in the past, but we prefer to not use `create or replace` instead if this is a new function just use `create`, and if this is replacing a function do a `drop` and `create`.